subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/ForkedTestClasspathFactory.java

```diff
@@ -140,33 +177,164 @@ private ImmutableList<URL> withImplementation(List<URL> additionalImplementation
+    private <T> ImmutableList<T> loadFromDistribution(List<TestFrameworkDistributionModule> moduleNames, Function<ClassPath, List<T>> extractor) {
         ImmutableList.Builder<T> outputFiles = ImmutableList.builder();
 
         if (LOGGER.isDebugEnabled() && !moduleNames.isEmpty()) {
             LOGGER.debug("Loaded additional modules from the Gradle distribution: " + Joiner.on(",").join(moduleNames));
         }
 
-        for (String module : moduleNames) {
-            ClassPath cp = moduleRegistry.getExternalModule(module).getImplementationClasspath();
+        for (TestFrameworkDistributionModule module : moduleNames) {
+            ClassPath cp = moduleRegistry.getExternalModule(module.getModuleName()).getImplementationClasspath();
             outputFiles.addAll(extractor.apply(cp));
         }
 
         return outputFiles.build();
     }
+
+    private static class AdditionalClasspath {
+
+        public final List<TestFrameworkDistributionModule> applicationClasspath;
+        public final List<TestFrameworkDistributionModule> applicationModulepath;
+        public final List<TestFrameworkDistributionModule> implementationClasspath;
+        public final List<TestFrameworkDistributionModule> implementationModulepath;
+
+        public AdditionalClasspath(
+            List<TestFrameworkDistributionModule> applicationClasspath,
+            List<TestFrameworkDistributionModule> applicationModulepath,
+            List<TestFrameworkDistributionModule> implementationClasspath,
+            List<TestFrameworkDistributionModule> implementationModulepath
+        ) {
+            this.applicationClasspath = applicationClasspath;
+            this.applicationModulepath = applicationModulepath;
+            this.implementationClasspath = implementationClasspath;
+            this.implementationModulepath = implementationModulepath;
+        }
+
+        public boolean isEmpty() {
+            return applicationClasspath.isEmpty() && applicationModulepath.isEmpty() &&
+                implementationClasspath.isEmpty() && implementationModulepath.isEmpty();
+        }
+    }
+
+    /**
+     * Filters the provided {@code unfiltered} additional classpath to only contain modules which are not already
+     * present in {@code classpath} and {@code modulepath}. This operates in a two-step process. First, it attempts
+     * to parse the names the jars in the provided classpath to quickly determine if the additional modules already
+     * exist. This is brittle and prone to errors, but much faster than the second step. If this step fails to filter
+     * all additional modules, the second step creates a {@link ClassLoader} based on the provided classpath and
+     * modulepath and attempts to discover the modules which are already present.
+     */
+    private AdditionalClasspath filterAdditionalClasspath(Iterable<? extends File> classpath, Iterable<? extends File> modulepath, AdditionalClasspath unfiltered) {
```

Comment:
I don't really think this is necessary to split up, for a couple reasons. 
1) Timeline. This is merging today
2) The complexity is not very high, at 350 lines this class not too long
3) The ClassDetector is not meant to be used externally. The only reason the functionality is pluggable and implemented by a factory is to allow the classloading to be mocked during unit testing, where we don't have actual jars to test against. 
4) The unit tests for this class are written and are able to fully test its functionality without extensive or complicated mocking

Reply:
If there were other _production_ filtering implementations, I would agree with you, but I don't think this is necessary to split up at this time.

Reply:
I don't object to merging this as-is to make 8.1, and I haven't thought about the details of the refactoring, but I still get a smell from this that it's too much for this class.

1) That seems most important, and I agree this shouldn't block merging.
2) It's not about the length, it's about the cohesion and the number of responsibilities of the class.
3) This is a valid reason to favor a single class, but I think it is outweighed by 2).  I think with some restructuring, you might not have to make these concessions for unit testing, which would also be a positive.
4) This would still be the case with a separate class - the test methods would just migrate.

Reply:
While this class may benefit from being split up, I think it is outweighed by the additional effort required to do so, mainly to update existing tests. In the current state, even though this class has two responsibilities, I do not believe it significantly contributes enough additional mental load in order to warrant splitting this up now. 

In any case, all of this code is used to support deprecated behavior, and thus will be removed in 9.0. I believe any additional work here is wasteful.